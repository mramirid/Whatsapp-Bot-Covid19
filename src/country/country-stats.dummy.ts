import type { CountryStats } from './country-stats.interface';

export const dummyCountryStats: CountryStats = {
  cases: 131_524_885,
  todayCases: 45_442_552,
  deaths: 3_352_999,
  todayDeaths: 823_993,
  recovered: 90_332,
  todayRecovered: 7223,
  active: 583,
  critical: 2,
  updatedAt: new Date('2023-05-14T01:24:30.408Z'),
};
