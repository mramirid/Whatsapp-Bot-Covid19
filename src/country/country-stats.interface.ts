export default interface CountryStats {
  cases: number;
  todayCases: number;
  deaths: number;
  todayDeaths: number;
  recovered: number;
  todayRecovered: number;
  active: number;
  critical: number;
  updatedAt: Date;
}
