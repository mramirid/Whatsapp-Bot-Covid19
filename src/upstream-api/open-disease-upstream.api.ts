import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { map } from 'rxjs';
import type { OpenDiseaseCountryStats } from './open-disease-country-stats.interface';
import { UpstreamAPI } from './upstream-api.abstract';

@Injectable()
export class OpenDiseaseUpstreamAPI extends UpstreamAPI {
  constructor(private readonly httpService: HttpService) {
    super();
  }

  getCountryStats() {
    return this.httpService
      .get<OpenDiseaseCountryStats>(
        'https://corona.lmao.ninja/v2/countries/ID?strict=true&query=',
      )
      .pipe(
        map((response) => response.data),
        map((stats) => ({
          cases: stats.cases,
          todayCases: stats.todayCases,
          deaths: stats.deaths,
          todayDeaths: stats.todayDeaths,
          recovered: stats.recovered,
          todayRecovered: stats.todayRecovered,
          active: stats.active,
          critical: stats.critical,
          updatedAt: new Date(stats.updated),
        })),
      );
  }
}
