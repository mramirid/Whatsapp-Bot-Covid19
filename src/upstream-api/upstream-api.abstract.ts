import type { Observable } from 'rxjs';

import type CountryStats from '../country/country-stats.interface';

export default abstract class UpstreamAPI {
  abstract getCountryStats(): Observable<CountryStats>;
}
