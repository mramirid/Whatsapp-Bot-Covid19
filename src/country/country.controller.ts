import { CacheInterceptor, CacheTTL } from '@nestjs/cache-manager';
import { Controller, Get, UseInterceptors } from '@nestjs/common';
import { CountryService } from './country.service';

@Controller('country')
export class CountryController {
  constructor(private readonly countryService: CountryService) {}

  @Get()
  @UseInterceptors(CacheInterceptor)
  @CacheTTL(86_400_000)
  getStatsMessage() {
    return this.countryService.getTodayStatsMessage();
  }
}
