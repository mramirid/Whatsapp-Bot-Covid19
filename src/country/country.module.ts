import { HttpModule } from '@nestjs/axios';
import { CacheModule } from '@nestjs/cache-manager';
import { Module } from '@nestjs/common';
import { OpenDiseaseUpstreamAPI } from 'src/open-disease/open-disease-upstream.api';
import { UpstreamAPI } from 'src/upstream-api.abstract';
import { CountryController } from './country.controller';
import { CountryService } from './country.service';

@Module({
  imports: [CacheModule.register(), HttpModule],
  controllers: [CountryController],
  providers: [
    CountryService,
    { provide: UpstreamAPI, useClass: OpenDiseaseUpstreamAPI },
  ],
})
export class CountryModule {}
