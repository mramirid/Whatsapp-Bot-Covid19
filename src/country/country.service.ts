import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Inject, Injectable } from '@nestjs/common';
import type { Cache } from 'cache-manager';
import { from, map, mergeMap, of, retry, tap } from 'rxjs';
import type { CountryStats } from '../upstream-api/interfaces/country-stats.interface';
import { UpstreamAPI } from '../upstream-api/upstream-api.abstract';

@Injectable()
export class CountryService {
  private readonly STATS_CACHE_KEY = 'COUNTRY_STATS_CACHE';

  constructor(
    private readonly upstreamAPI: UpstreamAPI,
    @Inject(CACHE_MANAGER) private readonly cacheManager: Cache,
  ) {}

  getTodayStatsMessage() {
    return this.getTodayStats().pipe(
      map((stats) => this.formatStats(stats)),
      map(
        (stats) =>
          '*Statistik COVID-19 di Indonesia*\n\n' +
          `Total: ${stats.cases} (+${stats.todayCases})\n` +
          `Sembuh: ${stats.recovered} (+${stats.todayRecovered})\n` +
          `Meninggal: ${stats.deaths} (+${stats.todayDeaths})\n` +
          `Dirawat: ${stats.active}\n` +
          `Kritis: ${stats.critical}\n\n` +
          `Tetap jaga kesehatan dan patuhi protokol kesehatan.\n\n` +
          `_Pembaharuan terakhir pada ${stats.updatedAt}._`,
      ),
    );
  }

  private getTodayStats() {
    return from(this.cacheManager.get<CountryStats>(this.STATS_CACHE_KEY)).pipe(
      mergeMap((cachedStats) => {
        if (cachedStats) {
          return of(cachedStats);
        }
        return this.upstreamAPI.getCountryStats().pipe(
          retry(2),
          tap((stats) => {
            this.cacheManager.set(this.STATS_CACHE_KEY, stats);
          }),
        );
      }),
    );
  }

  private formatStats(stats: CountryStats): Record<keyof CountryStats, string> {
    const numberFormatter = new Intl.NumberFormat('id-ID');
    const dateFormatter = new Intl.DateTimeFormat('id-ID', {
      dateStyle: 'medium',
      timeStyle: 'long',
    });

    return {
      cases: numberFormatter.format(stats.cases),
      todayCases: numberFormatter.format(stats.todayCases),
      recovered: numberFormatter.format(stats.recovered),
      todayRecovered: numberFormatter.format(stats.todayRecovered),
      deaths: numberFormatter.format(stats.deaths),
      todayDeaths: numberFormatter.format(stats.todayDeaths),
      active: numberFormatter.format(stats.active),
      critical: numberFormatter.format(stats.critical),
      updatedAt: dateFormatter.format(stats.updatedAt),
    };
  }
}
