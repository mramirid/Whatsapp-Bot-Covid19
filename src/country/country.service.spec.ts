import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Test } from '@nestjs/testing';
import { catchError, of, throwError } from 'rxjs';
import { firstValueFrom } from 'rxjs/internal/firstValueFrom';
import type { CountryStats } from '../upstream-api/interfaces/country-stats.interface';
import { UpstreamAPI } from '../upstream-api/upstream-api.abstract';
import { CountryService } from './country.service';

const countryStats: Readonly<CountryStats> = {
  cases: 131_524_885,
  todayCases: 45_442_552,
  deaths: 3_352_999,
  todayDeaths: 823_993,
  recovered: 90_332,
  todayRecovered: 7223,
  active: 583,
  critical: 2,
  updatedAt: new Date('2023-05-14T01:24:30.408Z'),
};

describe('CountryService - Unit Testing', () => {
  const getCacheMock = jest.fn();
  const setCacheMock = jest.fn();

  const getCountryStatsMock = jest.fn();

  let service: CountryService;

  let statsCacheKey: string;

  beforeEach(async () => {
    const cacheManagerMock = {
      get: getCacheMock,
      set: setCacheMock,
    };

    const upstreamAPIMock = {
      getCountryStats: getCountryStatsMock,
    };

    const module = await Test.createTestingModule({
      providers: [
        CountryService,
        { provide: CACHE_MANAGER, useValue: cacheManagerMock },
        { provide: UpstreamAPI, useValue: upstreamAPIMock },
      ],
    }).compile();

    service = module.get(CountryService);
    statsCacheKey = service['STATS_CACHE_KEY'];
  });

  afterEach(() => {
    getCacheMock.mockReset();
    setCacheMock.mockReset();
    getCountryStatsMock.mockReset();
  });

  describe('getTodayStatsMessage()', () => {
    it('should return the stats as a message with the correct format', async () => {
      getCacheMock.mockResolvedValueOnce(undefined);
      getCountryStatsMock.mockReturnValueOnce(of(countryStats));

      const gettingTodayStatsMessage = firstValueFrom(
        service.getTodayStatsMessage(),
      );

      await expect(gettingTodayStatsMessage).resolves.toMatchSnapshot();
    });
  });

  describe('getTodayStats()', () => {
    it('should get the stats from the cache if available', async () => {
      getCacheMock.mockResolvedValueOnce(countryStats);

      const gettingTodayStats = firstValueFrom(service['getTodayStats']());

      await expect(gettingTodayStats).resolves.toBe(countryStats);
      expect(getCacheMock).toHaveBeenCalledWith(statsCacheKey);
    });

    it('should fetch the stats from the upstream API if it is not available in the cache', async () => {
      getCacheMock.mockResolvedValueOnce(undefined);
      getCountryStatsMock.mockReturnValueOnce(of(countryStats));

      const gettingTodayStats = firstValueFrom(service['getTodayStats']());

      await expect(gettingTodayStats).resolves.toBe(countryStats);
      expect(getCountryStatsMock).toHaveBeenCalled();
    });

    it('should cache the stats after fetching the stats from the upstream API', async () => {
      getCacheMock.mockResolvedValueOnce(undefined);
      getCountryStatsMock.mockReturnValueOnce(of(countryStats));

      const gettingTodayStats = firstValueFrom(service['getTodayStats']());

      await expect(gettingTodayStats).resolves.toBe(countryStats);
      expect(setCacheMock).toHaveBeenCalledWith(statsCacheKey, countryStats);
    });

    it('should fetch the upstream API three times if it fails initially', (done) => {
      getCacheMock.mockResolvedValueOnce(undefined);
      getCountryStatsMock
        .mockReturnValueOnce(throwError(() => new Error()))
        .mockReturnValueOnce(throwError(() => new Error()))
        .mockReturnValueOnce(of(countryStats));

      let numFetches = 0;

      service['getTodayStats']()
        .pipe(
          catchError((_, caught) => {
            numFetches++;

            // We just incremented the counter here, so retry, do not throw.
            return caught;
          }),
        )
        .subscribe({
          complete: () => {
            expect(++numFetches).toBe(3);

            done();
          },
        });
    });

    it('should not cache the stats if upstream API requests fail', (done) => {
      getCacheMock.mockResolvedValueOnce(undefined);
      getCountryStatsMock.mockReturnValue(throwError(() => new Error()));

      service['getTodayStats']().subscribe({
        error: () => {
          expect(setCacheMock).not.toBeCalled();
          done();
        },
      });
    });
  });

  describe('formatStats()', () => {
    const numberFormatter = new Intl.NumberFormat('id-ID');
    const dateFormatter = new Intl.DateTimeFormat('id-ID', {
      dateStyle: 'medium',
      timeStyle: 'long',
    });

    it('should return formatted stats localized in Indonesian', () => {
      const formattedStats = service['formatStats'](countryStats);

      expect(formattedStats).toEqual({
        cases: numberFormatter.format(countryStats.cases),
        todayCases: numberFormatter.format(countryStats.todayCases),
        recovered: numberFormatter.format(countryStats.recovered),
        todayRecovered: numberFormatter.format(countryStats.todayRecovered),
        deaths: numberFormatter.format(countryStats.deaths),
        todayDeaths: numberFormatter.format(countryStats.todayDeaths),
        active: numberFormatter.format(countryStats.active),
        critical: numberFormatter.format(countryStats.critical),
        updatedAt: dateFormatter.format(countryStats.updatedAt),
      });
    });
  });
});
