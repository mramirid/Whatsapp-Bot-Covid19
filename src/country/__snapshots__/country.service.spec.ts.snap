// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryService - Unit Testing getTodayStatsMessage() should return the stats as a message with the correct format 1`] = `
"*Statistik COVID-19 di Indonesia*

Total: 131.524.885 (+45.442.552)
Sembuh: 90.332 (+7.223)
Meninggal: 3.352.999 (+823.993)
Dirawat: 583
Kritis: 2

Tetap jaga kesehatan dan patuhi protokol kesehatan.

_Pembaharuan terakhir pada 14 Mei 2023, 08.24.30 WIB._"
`;
